content/en/platform/corda/5.0/developing-applications/packaging/code-signing/custom-cert.md

```diff
@@ -0,0 +1,58 @@
+---
+date: '2023-05-18'
+version: 'Corda 5.0'
+title: "Signing CPKs and CPBs with a Custom Certificate"
+menu:
+  corda5:
+    parent: corda5-develop-packaging-code-signing
+    identifier: corda5-develop-packaging-code-signing-custom-cert
+    weight: 1000
+section_menu: corda5
+---
+
+# Signing CPKs and CPBs with a Custom Certificate
+
+As described in [Build a CPB](../cpb.md) and [Build a CPK](../cpk.md), the Gradle plugin uses, by default,
+a development certificate to sign a CPB or CPK package.
+
+However, you can configure the plugin to use a custom certificate using the `cordapp` `signing` section of the Gradle plugin:
+
+```
+cordapp {
+    ...
+
+    signing {
+        enabled = (true | false)
+        options {
+            alias = '??'
```

Comment:
@josephzunigadaly , do you know if/where we have documentation about all of these configuration options?

I agree with @yarivus , `??` is a bit odd here (I think I copied it from the plugin repo readme).

Reply:
Alias is the key name within the keystore. It is a term that tools like keytool and jarsigner use.

Here is an example of signing configured to work within the CSDE template. I wrote this intending it to be included in the docs but using this causes the CPI build to fail because we don't trust the private keys inside the keystore. CORE-7275 should fix that.

```groovy
signing {
    options {
        alias = "my-signing-key"
        storePassword = "keystore password"
        keyStore = layout.projectDirectory.dir("../workspace").file("signingkeys.pfx").toString()
        storeType = "PKCS12"
    }
}
```

Reply:
We don't have existing docs. I'll try to explain the parameters here:

* alias - name/identifier of the key within the keystore
* storePassword - password for accessing keystore
* keyStore - path of keystore file
* storeType- 'PKCS12' or 'JKS', different formats for storing the key, we prefer PKCS12
* keyPassword - password for accessing the key within the keystore, not all keystores support per-key passwords, we don't need it in any of our examples
* signatureFileName - controls names of the signature files within the jar file, defaults to match alias
* verbose - should we output extra information on the signing process - useful for debugging failures
* strict - Not sure - Needs further investigation
* internalSF - Not sure - Needs further investigation
* sectionsOnly - Not sure - Needs further investigation
* lazy - Not sure - Needs further investigation
* maxMemory - Guessing this is a memory limit but need to check
* preserveLastModified - Appears to be an option to preserve timestamps, so they are not changed. Need to dig into which files this affects. This might be useful if you have set all timestamps to a standard time so that repeat builds of the same code produce the same output jar.
* tsaUrl - timestamp server url. This is a system where a third party apply a signature to prove the time it was signed. The idea here is that signatures will expire but we want to keep running code beyond the expiry. The timestamp allows logic like if it was signed while the cert was valid (between start and end dates of the cert), trust the signature. The third party is needed because they are independent. The owner of the cert may have interest in applying a false timestamp.
* tsaCert - Not sure - Needs further investigation, related to TSA
* tsaProxyHost - TSA is often an external service on the internet and may need proxy server configured for access, this lets you configure the proxy server host
* tsaProxyPort - configure proxy server port
* executable - Appears to allow changing the jarsigner command in use
* force - Not sure - Needs further investigation
* signatureAlgorithm - Not sure - Needs further investigation
* digestAlgorithm - Not sure - Needs further investigation
* tsaDigestAlgorithm - Not sure - Needs further investigation

Reply:
I had a look at the plugin code for the options I did not understand. I found that the plugin uses the Ant signjar task for signing. So the Ant docs describe the options:

https://ant.apache.org/manual/Tasks/signjar.html